{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5957a1dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from qiime2 import Artifact\n",
    "from qiime2.plugin.testing import TestPluginBase\n",
    "from pandas import DataFrame as DataFrame\n",
    "\n",
    "from q2_types_genomics.eggnog import NOG, ArbitraryHeaderTSVFmt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7063460d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "69fdf775",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = 'sampleannotations.txt'\n",
    "\n",
    "filepath = os.path.join(\"./tests/data/\", filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "df953e5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ff = ArbitraryHeaderTSVFmt(filepath, 'r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5f308ae8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#query_name</th>\n",
       "      <th>seed_eggNOG_ortholog</th>\n",
       "      <th>seed_ortholog_evalue</th>\n",
       "      <th>seed_ortholog_score</th>\n",
       "      <th>eggNOG OGs</th>\n",
       "      <th>narr_og_name</th>\n",
       "      <th>narr_og_cat</th>\n",
       "      <th>narr_og_desc</th>\n",
       "      <th>best_og_name</th>\n",
       "      <th>best_og_cat</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k129_5480_1</td>\n",
       "      <td>856793.MICA_1589</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>982.2</td>\n",
       "      <td>COG0458@1|root,COG0458@2|Bacteria,1MUDZ@1224|P...</td>\n",
       "      <td>4BPM5@82117|unclassified Alphaproteobacteria</td>\n",
       "      <td>F</td>\n",
       "      <td>PFAM Carbamoyl-phosphate synthase L chain, ATP...</td>\n",
       "      <td>2TQZU@28211|Alphaproteobacteria</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K01955</td>\n",
       "      <td>ko00240,ko00250,ko01100,map00240,map00250,map0...</td>\n",
       "      <td>M00051</td>\n",
       "      <td>R00256,R00575,R01395,R10948,R10949</td>\n",
       "      <td>RC00002,RC00010,RC00043,RC02750,RC02798,RC03314</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ATP-grasp,ATP-grasp_3,ATP-grasp_4,CPSase_L_D2,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k129_5480_2</td>\n",
       "      <td>1469613.JT55_09850</td>\n",
       "      <td>1.600000e-67</td>\n",
       "      <td>209.1</td>\n",
       "      <td>COG0782@1|root,COG0782@2|Bacteria,1RCXW@1224|P...</td>\n",
       "      <td>3FD6X@34008|Rhodovulum</td>\n",
       "      <td>K</td>\n",
       "      <td>Necessary for efficient RNA polymerase transcr...</td>\n",
       "      <td>2U5JU@28211|Alphaproteobacteria</td>\n",
       "      <td>K</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K03624</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko03021</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GreA_GreB,GreA_GreB_N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k129_5480_3</td>\n",
       "      <td>985867.AEWF01000001_gene1800</td>\n",
       "      <td>5.800000e-61</td>\n",
       "      <td>192.6</td>\n",
       "      <td>COG1522@1|root,COG1522@2|Bacteria,1MXVU@1224|P...</td>\n",
       "      <td>47FT9@766|Rickettsiales</td>\n",
       "      <td>K</td>\n",
       "      <td>helix_turn_helix ASNC type</td>\n",
       "      <td>47FT9@766|Rickettsiales</td>\n",
       "      <td>K</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>AsnC_trans_reg,HTH_24,HTH_AsnC-type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k129_5480_4</td>\n",
       "      <td>856793.MICA_1598</td>\n",
       "      <td>4.600000e-63</td>\n",
       "      <td>209.1</td>\n",
       "      <td>COG3660@1|root,COG3660@2|Bacteria,1R70U@1224|P...</td>\n",
       "      <td>4BTDH@82117|unclassified Alphaproteobacteria</td>\n",
       "      <td>M</td>\n",
       "      <td>Mitochondrial fission ELM1</td>\n",
       "      <td>2TQWV@28211|Alphaproteobacteria</td>\n",
       "      <td>M</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K07276</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Mito_fiss_Elm1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k129_5480_5</td>\n",
       "      <td>1528106.JRJE01000008_gene1538</td>\n",
       "      <td>1.900000e-160</td>\n",
       "      <td>457.6</td>\n",
       "      <td>COG0492@1|root,COG0492@2|Bacteria,1MV15@1224|P...</td>\n",
       "      <td>2JPS7@204441|Rhodospirillales</td>\n",
       "      <td>C</td>\n",
       "      <td>Belongs to the class-II pyridine nucleotide-di...</td>\n",
       "      <td>2JPS7@204441|Rhodospirillales</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>ko:K00384</td>\n",
       "      <td>ko00450,map00450</td>\n",
       "      <td>-</td>\n",
       "      <td>R02016,R03596,R09372</td>\n",
       "      <td>RC00013,RC02518,RC02873</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>FAD_binding_2,FMO-like,GIDA,K_oxygenase,Pyr_re...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1370</th>\n",
       "      <td>k129_5480_1788</td>\n",
       "      <td>1294273.roselon_02816</td>\n",
       "      <td>1.200000e-22</td>\n",
       "      <td>88.2</td>\n",
       "      <td>COG0401@1|root,COG0401@2|Bacteria,1N7K3@1224|P...</td>\n",
       "      <td>2UFGT@28211|Alphaproteobacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>homolog of Blt101</td>\n",
       "      <td>2UFGT@28211|Alphaproteobacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pmp3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1371</th>\n",
       "      <td>k129_5480_1789</td>\n",
       "      <td>501479.ACNW01000063_gene2789</td>\n",
       "      <td>3.400000e-23</td>\n",
       "      <td>90.1</td>\n",
       "      <td>COG3237@1|root,COG3237@2|Bacteria,1N6X4@1224|P...</td>\n",
       "      <td>2UF5Q@28211|Alphaproteobacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>Belongs to the UPF0337 (CsbD) family</td>\n",
       "      <td>2UF5Q@28211|Alphaproteobacteria</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CsbD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1372</th>\n",
       "      <td># 1372 queries scanned</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1373</th>\n",
       "      <td># Total time (seconds): 7.36770224571228</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1374</th>\n",
       "      <td># Rate: 186.22 q/s</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1375 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   #query_name           seed_eggNOG_ortholog  \\\n",
       "0                                  k129_5480_1               856793.MICA_1589   \n",
       "1                                  k129_5480_2             1469613.JT55_09850   \n",
       "2                                  k129_5480_3   985867.AEWF01000001_gene1800   \n",
       "3                                  k129_5480_4               856793.MICA_1598   \n",
       "4                                  k129_5480_5  1528106.JRJE01000008_gene1538   \n",
       "...                                        ...                            ...   \n",
       "1370                            k129_5480_1788          1294273.roselon_02816   \n",
       "1371                            k129_5480_1789   501479.ACNW01000063_gene2789   \n",
       "1372                    # 1372 queries scanned                            NaN   \n",
       "1373  # Total time (seconds): 7.36770224571228                            NaN   \n",
       "1374                        # Rate: 186.22 q/s                            NaN   \n",
       "\n",
       "      seed_ortholog_evalue  seed_ortholog_score  \\\n",
       "0             0.000000e+00                982.2   \n",
       "1             1.600000e-67                209.1   \n",
       "2             5.800000e-61                192.6   \n",
       "3             4.600000e-63                209.1   \n",
       "4            1.900000e-160                457.6   \n",
       "...                    ...                  ...   \n",
       "1370          1.200000e-22                 88.2   \n",
       "1371          3.400000e-23                 90.1   \n",
       "1372                   NaN                  NaN   \n",
       "1373                   NaN                  NaN   \n",
       "1374                   NaN                  NaN   \n",
       "\n",
       "                                             eggNOG OGs  \\\n",
       "0     COG0458@1|root,COG0458@2|Bacteria,1MUDZ@1224|P...   \n",
       "1     COG0782@1|root,COG0782@2|Bacteria,1RCXW@1224|P...   \n",
       "2     COG1522@1|root,COG1522@2|Bacteria,1MXVU@1224|P...   \n",
       "3     COG3660@1|root,COG3660@2|Bacteria,1R70U@1224|P...   \n",
       "4     COG0492@1|root,COG0492@2|Bacteria,1MV15@1224|P...   \n",
       "...                                                 ...   \n",
       "1370  COG0401@1|root,COG0401@2|Bacteria,1N7K3@1224|P...   \n",
       "1371  COG3237@1|root,COG3237@2|Bacteria,1N6X4@1224|P...   \n",
       "1372                                                NaN   \n",
       "1373                                                NaN   \n",
       "1374                                                NaN   \n",
       "\n",
       "                                      narr_og_name narr_og_cat  \\\n",
       "0     4BPM5@82117|unclassified Alphaproteobacteria           F   \n",
       "1                           3FD6X@34008|Rhodovulum           K   \n",
       "2                          47FT9@766|Rickettsiales           K   \n",
       "3     4BTDH@82117|unclassified Alphaproteobacteria           M   \n",
       "4                    2JPS7@204441|Rhodospirillales           C   \n",
       "...                                            ...         ...   \n",
       "1370               2UFGT@28211|Alphaproteobacteria           S   \n",
       "1371               2UF5Q@28211|Alphaproteobacteria           S   \n",
       "1372                                           NaN         NaN   \n",
       "1373                                           NaN         NaN   \n",
       "1374                                           NaN         NaN   \n",
       "\n",
       "                                           narr_og_desc  \\\n",
       "0     PFAM Carbamoyl-phosphate synthase L chain, ATP...   \n",
       "1     Necessary for efficient RNA polymerase transcr...   \n",
       "2                            helix_turn_helix ASNC type   \n",
       "3                            Mitochondrial fission ELM1   \n",
       "4     Belongs to the class-II pyridine nucleotide-di...   \n",
       "...                                                 ...   \n",
       "1370                                  homolog of Blt101   \n",
       "1371               Belongs to the UPF0337 (CsbD) family   \n",
       "1372                                                NaN   \n",
       "1373                                                NaN   \n",
       "1374                                                NaN   \n",
       "\n",
       "                         best_og_name best_og_cat  ...    KEGG_ko  \\\n",
       "0     2TQZU@28211|Alphaproteobacteria           F  ...  ko:K01955   \n",
       "1     2U5JU@28211|Alphaproteobacteria           K  ...  ko:K03624   \n",
       "2             47FT9@766|Rickettsiales           K  ...          -   \n",
       "3     2TQWV@28211|Alphaproteobacteria           M  ...  ko:K07276   \n",
       "4       2JPS7@204441|Rhodospirillales           C  ...  ko:K00384   \n",
       "...                               ...         ...  ...        ...   \n",
       "1370  2UFGT@28211|Alphaproteobacteria           S  ...          -   \n",
       "1371  2UF5Q@28211|Alphaproteobacteria           S  ...          -   \n",
       "1372                              NaN         NaN  ...        NaN   \n",
       "1373                              NaN         NaN  ...        NaN   \n",
       "1374                              NaN         NaN  ...        NaN   \n",
       "\n",
       "                                           KEGG_Pathway KEGG_Module  \\\n",
       "0     ko00240,ko00250,ko01100,map00240,map00250,map0...      M00051   \n",
       "1                                                     -           -   \n",
       "2                                                     -           -   \n",
       "3                                                     -           -   \n",
       "4                                      ko00450,map00450           -   \n",
       "...                                                 ...         ...   \n",
       "1370                                                  -           -   \n",
       "1371                                                  -           -   \n",
       "1372                                                NaN         NaN   \n",
       "1373                                                NaN         NaN   \n",
       "1374                                                NaN         NaN   \n",
       "\n",
       "                           KEGG_Reaction  \\\n",
       "0     R00256,R00575,R01395,R10948,R10949   \n",
       "1                                      -   \n",
       "2                                      -   \n",
       "3                                      -   \n",
       "4                   R02016,R03596,R09372   \n",
       "...                                  ...   \n",
       "1370                                   -   \n",
       "1371                                   -   \n",
       "1372                                 NaN   \n",
       "1373                                 NaN   \n",
       "1374                                 NaN   \n",
       "\n",
       "                                          KEGG_rclass  \\\n",
       "0     RC00002,RC00010,RC00043,RC02750,RC02798,RC03314   \n",
       "1                                                   -   \n",
       "2                                                   -   \n",
       "3                                                   -   \n",
       "4                             RC00013,RC02518,RC02873   \n",
       "...                                               ...   \n",
       "1370                                                -   \n",
       "1371                                                -   \n",
       "1372                                              NaN   \n",
       "1373                                              NaN   \n",
       "1374                                              NaN   \n",
       "\n",
       "                                BRITE KEGG_TC CAZy BiGG_Reaction  \\\n",
       "0     ko00000,ko00001,ko00002,ko01000       -    -             -   \n",
       "1                     ko00000,ko03021       -    -             -   \n",
       "2                                   -       -    -             -   \n",
       "3                             ko00000       -    -             -   \n",
       "4             ko00000,ko00001,ko01000       -    -             -   \n",
       "...                               ...     ...  ...           ...   \n",
       "1370                                -       -    -             -   \n",
       "1371                                -       -    -             -   \n",
       "1372                              NaN     NaN  NaN           NaN   \n",
       "1373                              NaN     NaN  NaN           NaN   \n",
       "1374                              NaN     NaN  NaN           NaN   \n",
       "\n",
       "                                                  PFAMs  \n",
       "0     ATP-grasp,ATP-grasp_3,ATP-grasp_4,CPSase_L_D2,...  \n",
       "1                                 GreA_GreB,GreA_GreB_N  \n",
       "2                   AsnC_trans_reg,HTH_24,HTH_AsnC-type  \n",
       "3                                        Mito_fiss_Elm1  \n",
       "4     FAD_binding_2,FMO-like,GIDA,K_oxygenase,Pyr_re...  \n",
       "...                                                 ...  \n",
       "1370                                               Pmp3  \n",
       "1371                                               CsbD  \n",
       "1372                                                NaN  \n",
       "1373                                                NaN  \n",
       "1374                                                NaN  \n",
       "\n",
       "[1375 rows x 24 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ff.view(DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "55064b2e",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "No correct separator detected in input file on line: 1",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [5]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m test_file \u001b[38;5;241m=\u001b[39m \u001b[43mArtifact\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mimport_data\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mFeatureData[NOG]\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfilepath\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/qiime2/qiime2/qiime2/sdk/result.py:278\u001b[0m, in \u001b[0;36mArtifact.import_data\u001b[0;34m(cls, type, view, view_type)\u001b[0m\n\u001b[1;32m    275\u001b[0m     format_ \u001b[38;5;241m=\u001b[39m view_type\n\u001b[1;32m    277\u001b[0m provenance_capture \u001b[38;5;241m=\u001b[39m archive\u001b[38;5;241m.\u001b[39mImportProvenanceCapture(format_, md5sums)\n\u001b[0;32m--> 278\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mcls\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_from_view\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtype_\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mview\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mview_type\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprovenance_capture\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    279\u001b[0m \u001b[43m                      \u001b[49m\u001b[43mvalidate_level\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmax\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/qiime2/qiime2/qiime2/sdk/result.py:306\u001b[0m, in \u001b[0;36mArtifact._from_view\u001b[0;34m(cls, type, view, view_type, provenance_capture, validate_level)\u001b[0m\n\u001b[1;32m    303\u001b[0m recorder \u001b[38;5;241m=\u001b[39m provenance_capture\u001b[38;5;241m.\u001b[39mtransformation_recorder(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mreturn\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    304\u001b[0m transformation \u001b[38;5;241m=\u001b[39m from_type\u001b[38;5;241m.\u001b[39mmake_transformation(to_type,\n\u001b[1;32m    305\u001b[0m                                                recorder\u001b[38;5;241m=\u001b[39mrecorder)\n\u001b[0;32m--> 306\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mtransformation\u001b[49m\u001b[43m(\u001b[49m\u001b[43mview\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalidate_level\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    308\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m type_raw \u001b[38;5;129;01min\u001b[39;00m pm\u001b[38;5;241m.\u001b[39mvalidators:\n\u001b[1;32m    309\u001b[0m     validation_object \u001b[38;5;241m=\u001b[39m pm\u001b[38;5;241m.\u001b[39mvalidators[\u001b[38;5;28mtype\u001b[39m]\n",
      "File \u001b[0;32m~/work/qiime2/qiime2/qiime2/core/transform.py:68\u001b[0m, in \u001b[0;36mModelType.make_transformation.<locals>.transformation\u001b[0;34m(view, validate_level)\u001b[0m\n\u001b[1;32m     66\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mtransformation\u001b[39m(view, validate_level\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmin\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m     67\u001b[0m     view \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcoerce_view(view)\n\u001b[0;32m---> 68\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mview\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalidate_level\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     70\u001b[0m     new_view \u001b[38;5;241m=\u001b[39m transformer(view)\n\u001b[1;32m     72\u001b[0m     new_view \u001b[38;5;241m=\u001b[39m other\u001b[38;5;241m.\u001b[39mcoerce_view(new_view)\n",
      "File \u001b[0;32m~/work/qiime2/qiime2/qiime2/core/transform.py:143\u001b[0m, in \u001b[0;36mFormatType.validate\u001b[0;34m(self, view, level)\u001b[0m\n\u001b[1;32m    140\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m is not an instance of \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    141\u001b[0m                     \u001b[38;5;241m%\u001b[39m (view, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_view_type))\n\u001b[1;32m    142\u001b[0m \u001b[38;5;66;03m# Formats have a validate method, so defer to it\u001b[39;00m\n\u001b[0;32m--> 143\u001b[0m \u001b[43mview\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlevel\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/qiime2/qiime2/qiime2/plugin/model/file_format.py:26\u001b[0m, in \u001b[0;36m_FileFormat.validate\u001b[0;34m(self, level)\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_validate_\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m     25\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 26\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlevel\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     27\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m ValidationError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m     28\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m ValidationError(\n\u001b[1;32m     29\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m is not a(n) \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m file:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     30\u001b[0m             \u001b[38;5;241m%\u001b[39m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpath, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m, \u001b[38;5;28mstr\u001b[39m(e))\n\u001b[1;32m     31\u001b[0m             ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n",
      "File \u001b[0;32m~/work/bokulich-lab/q2-types-genomics/q2_types_genomics/eggnog/_format.py:34\u001b[0m, in \u001b[0;36mEggnogAnnotationFmt._validate_\u001b[0;34m(self, level)\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_validate_\u001b[39m(\u001b[38;5;28mself\u001b[39m, level):\n\u001b[0;32m---> 34\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_check_seperator\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmin\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m5\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmax\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m}\u001b[49m\u001b[43m[\u001b[49m\u001b[43mlevel\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/work/bokulich-lab/q2-types-genomics/q2_types_genomics/eggnog/_format.py:26\u001b[0m, in \u001b[0;36mEggnogAnnotationFmt._check_seperator\u001b[0;34m(self, level)\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, line \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(fh, \u001b[38;5;241m1\u001b[39m):\n\u001b[1;32m     25\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m re\u001b[38;5;241m.\u001b[39msearch(\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124mt\u001b[39m\u001b[38;5;124m'\u001b[39m, line):\n\u001b[0;32m---> 26\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNo correct separator detected in input \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     27\u001b[0m                          \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfile on line: \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(i))\n\u001b[1;32m     29\u001b[0m     re\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124mt\u001b[39m\u001b[38;5;124m'\u001b[39m, line)\n\u001b[1;32m     30\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m i \u001b[38;5;241m==\u001b[39m level:\n",
      "\u001b[0;31mValueError\u001b[0m: No correct separator detected in input file on line: 1"
     ]
    }
   ],
   "source": [
    "test_file = Artifact.import_data('FeatureData[NOG]', filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f1900e9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
